lang/macro/src/codegen/dispatch.rs

```diff
@@ -70,6 +77,238 @@ impl GenerateCode for Dispatch<'_> {
 }
 
 impl Dispatch<'_> {
+    fn generate_dispatch_variant_ident(
+        &self,
+        message: &ir::Function,
+        prefix: &str,
+    ) -> Ident {
+        let selector = message
+            .selector()
+            .expect("encountered a non-constructor function");
+        let selector_bytes = selector.as_bytes();
+        quote::format_ident!(
+            "__{}_0x{:02X}{:02X}{:02X}{:02X}",
+            prefix,
+            selector_bytes[0],
+            selector_bytes[1],
+            selector_bytes[2],
+            selector_bytes[3]
+        )
+    }
+
+    fn generate_dispatch_variant_decode(
+        &self,
+        message: &ir::Function,
+        prefix: &str,
+    ) -> TokenStream2 {
+        let selector_bytes = *message
+            .selector()
+            .expect("encountered a non-message function")
+            .as_bytes();
+        let s0 = selector_bytes[0];
+        let s1 = selector_bytes[1];
+        let s2 = selector_bytes[2];
+        let s3 = selector_bytes[3];
+        let variant_ident = self.generate_dispatch_variant_ident(message, prefix);
+        let variant_types = message.sig.inputs().map(|arg| &arg.ty);
+        quote! {
+            [#s0, #s1, #s2, #s3] => {
+                Ok(Self::#variant_ident(
+                    #(
+                        <#variant_types as ::scale::Decode>::decode(input)?
+                    ),*
+                ))
+            }
+        }
+    }
+
+    fn generate_dispatch_variant_arm(
+        &self,
+        message: &ir::Function,
+        prefix: &str,
+    ) -> TokenStream2 {
+        let input_types = message.sig.inputs().map(|arg| &arg.ty);
+        let variant_ident = self.generate_dispatch_variant_ident(message, prefix);
+        quote! {
+            #variant_ident(#(#input_types),*)
+        }
+    }
+
+    fn generate_message_dispatch_enum(&self) -> TokenStream2 {
+        let storage_ident = &self.contract.storage.ident;
+        let message_variants = self
+            .contract
+            .functions
+            .iter()
+            .filter(|function| function.is_message())
```

Comment:
This code block filtering message functions is repeated three times here, could be done once.